Added ShuffleWithinSegments

Erlemar opened the pull request:
Shuffling tokens within segments.

    The token sequence is split into segments of the same label; for each segment, a decision is made randomly: to shuffle it or not.
    In case of a positive decision, the tokens within each segment are shuffled.

The source: https://arxiv.org/abs/2010.11683

Erlemar commented:
SuspectingParaphraser doesn't pass the tests. What should I do about it.

sebastianGehrmann commented on transformations/shuffle_within_segments/test.json:
```diff
@@ -0,0 +1,71 @@
+{
```
The method signature expects Lists but your test cases use strings.

Erlemar replied:
The problem is that tests for my transformation (ShuffleWithinSegments) pass successfully, tests fail for a different transformation (SuspectingParaphraser).
As for the signature, I have followed the example of some other TEXT_TAGGING transformers like https://github.com/GEM-benchmark/NL-Augmenter/tree/main/transformations/longer_location_ner

Erlemar replied:
The tests work now.

sebastianGehrmann reviewed (changes requested):
I suspect your tests fails since your test cases do not conform to the expect data format (list vs. string). I can take another look when this is fixed.

marco-digio commented:
Hi @Erlemar 
I was assigned as one of your reviewers. 
Thanks for this contribution! 
I this that this transformation is very useful and clear. 

Here's my general review:
Correctness: All checks have failed, please fix this by looking at other PRs
Interface: I think you have chosen the correct interface(s).
Applicable Tasks & Keywords: I think you have chosen the relevant task. Please add keywords as explained [here](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)
Specificity: The change is general.
Novelty: This transformation is not yet implemented in NL-Augmenter.
Adding New Libraries: No new libraries appear to be added, and there is no requirements.txt file.
Description: The README is clear and explains what the transformation is aiming to do.
Data and code source: References clearly indicated in README
Test Cases: 6 test cases have been added.
Evaluating Robustness: No robustness evaluation has been performed.
Languages other than English: Yes, but it is required as input a TAG sequence

Erlemar commented:
Hello @marco-digio 

Thank you for your review!

Correctness:

The problem is that tests for my transformation (ShuffleWithinSegments) pass successfully, tests fail for a different transformation (SuspectingParaphraser). I'm not sure what should I do about it.

Applicable Tasks & Keywords: I have added the keywords, but I'm not sure that I have selected all of them correctly. Here is my reasoning:
"lexical" - the transformation substitutes different words.
"word-order" -  the order of the words is changed.
"rule-based" - the transformation uses heuristics.
"unnaturally-written", "unnatural-sounding" - I'm not sure about these keywords: this transformation randomly shuffles words within separate entities, as a result, the order of the words could be unnatural. This perturbed order could sound unnatural too. This is why I chose both these keywords.
"possible-meaning-preserving" - due to the changed word order, the meaning of the sentence could change.
"high-precision" - the transformation generates what it intends to generate.
"low-coverage" - the transformation works only for TAG sequences.
"low-generations" - The transformation generates few outputs for a single input.

Evaluating Robustness: I have added the results of the evaluation check.

Erlemar commented:
I have pushed the changes from https://github.com/GEM-benchmark/NL-Augmenter/pull/288 now the tests should pass.

samsontmr reviewed (commented):
Hi @Erlemar! It seems tests are still failing

Erlemar commented:
Hi @samsontmr @marco-digio the tests have passsed successfully.

Erlemar reviewed (commented):
The tests work now.

samsontmr commented on transformations/shuffle_within_segments/README.md:
```diff
@@ -0,0 +1,40 @@
+
+The average accuracy on its perturbed set = 70.13742999132477
+
+## Previous Work
```
Could you replace this bibtex with the one from [the ACL anthology](https://aclanthology.org/2020.coling-main.343)?

Erlemar replied:
I have replaced this bibtex.

samsontmr commented on transformations/shuffle_within_segments/README.md:
```diff
@@ -0,0 +1,40 @@
+This would help augment data for a NER task by keeping the labels still aligned.
+
+## What are the limitations of this transformation?
+It is possible that shuffling could change the meaning of the sentence or have negative effect for NLU tasks.
```
Another effect is that changing the word order could change the entity's "meaning"

Erlemar replied:
Updated the description.

samsontmr reviewed (commented):
Looks to mostly be in good shape!

samsontmr commented on transformations/shuffle_within_segments/README.md:
```diff
@@ -13,7 +13,7 @@ text generation, and most importantly, a tagging task.
 
 ## What are the limitations of this transformation?
-It is possible that shuffling could change the meaning of the sentence or have negative effect for NLU tasks.
+It is possible that shuffling could change the meaning of the sentence, the meaning the entity itself or have negative effect for NLU tasks.
```
the meaning **_of_** the entity itself or have **_a_** negative effect **_on_** NLU tasks

Erlemar replied:
I have fixed the typos.

samsontmr commented on transformations/shuffle_within_segments/transformation.py:
```diff
@@ -0,0 +1,70 @@
+    def generate(
+            self, token_sequence: List[str], tag_sequence: List[str]
+    ) -> List[Tuple[List[str], List[str]]]:
+        random.seed(self.seed)
```
I believe this is no longer needed

Erlemar replied:
I have removed it now.

kaustubhdhole commented:
Thanks @Erlemar for the hard work and everyone for their comments.
